### ISSUE TITLE
Prevent removal of apiVersions while they are still used

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#8566]
### What would you like to be added (User Story)?

As a user it would be nice if Cluster API would block removal of apiVersions of CRDs as long as they are still used.

### Detailed Description

Today it is possible that during a Cluster API upgrade (e.g. when bumping an infra provider) an apiVersion is removed. If this apiVersion was still referenced in other resources, they won't reconcile anymore.

Affected resources:
* ClusterClass
* Cluster, MD, MS, Machine, ...
  * Refs are probably not affected as they don't actually use the apiVersion from the object spec (they calculate the version they use based on the contract label on the CRD).
    * Would be good to double check this again.
  * OwnerRefs of these resources are relevant though.

A possible solution is to implement a webhook which validates CRD updates. In this webhook we could check if the update on the CRD removes an apiVersion which is still referenced in relevant resources (at least ClusterClass)

Note: This webhook should probably have failurePolicy: Ignore to avoid deadlocks if Cluster API is not working for some reason.

### Anything else you would like to add?

Somewhat related: https://github.com/kubernetes-sigs/cluster-api/issues/6539

### Label(s) to be applied

/kind feature
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by killianmuldoon** [comment]
One clarification - how does this relate to

>  Ensure clusterctl upgrade warns users if the upgrade would break them (e.g. v0.3 => v1.x (version without v1alpha3 apiVersion))
>    This should already be covered by our CRD migration code, but let's verify it.
>    This might only become relevant with v1.6 when we actually remove the apiVersion from the CRDs

From https://github.com/kubernetes-sigs/cluster-api/issues/8038

**Comment by fabriziopandini** [comment]
This is a nice check to have, but it comes with a couple of problems to be addressed, top of mind:
1. Where to do the check (webhook or clusterctl)
2. How to encode the knowledge of where references exist

For 1, it seems to me that the simplest way forward is to rely on clusterctl, because it already has all the discovery logic that is required for performing such a check (I'm assuming the discovery logic that exists in clusterctl move can be re-used for this task).

For 2, we can start thinking of a solution that focuses on references existing in core CAPI types, so we cover 90% of the issue. Then, in a follow-up iteration, we can extend this for references existing in provider types.

/help

**Comment by sbueringer** [comment]
> This is a nice check to have, but it comes with a couple of problems to be addressed, top of mind:
> 
> 1. Where to do the check (webhook or clusterctl)
> 2. How to encode the knowledge of where references exist
> 
> For 1, it seems to me that the simplest way forward is to rely on clusterctl, because it already has all the discovery logic that is required for performing such a check (I'm assuming the discovery logic that exists in clusterctl move can be re-used for this task).
> 
> For 2, we can start thinking of a solution that focuses on references existing in core CAPI types, so we cover 90% of the issue. Then, in a follow-up iteration, we can extend this for references existing in provider types.


1. I would do it in the webhook so everyone profits not only clusterctl users. (we don't need the discovery logic if we only implement it for core resources)
2. I think we don't have to encode that knowledge. I would implement a webhook for the core CAPI resources. Specifically I think only ClusterClass, in all other resources we automatically bump the apiVersion based on contract (which is also what providers should do). ClusterClass should be the only case where we intentionally use specific apiVersions.

**Comment by ykakarap** [comment]
So a validating webhook for `kind: CustomResourceDefinition` that checks that when a CRD whose `.spec.group` one of `controlplane.cluster.x-k8s.io` or `bootstrap.cluster.x-k8s.io` or `infrastructure.cluster.x-k8s.io` is updated does not drop a version that is used by at least one ClusterClass. Is my understanding correct?

**Comment by sbueringer** [comment]
Pretty much. I think we don't have to filter early by group. We can just take the group+ kind etc and check if it occurs in a ClusterClass 

This way we can also support CRDs in other groups. Afaik there is no contract that a provider has to use one of those groups

**Comment by sbueringer** [comment]
I think that's okay. It's just important that we use `failurePolicy: Ignore`. This ensures that the cluster is not broken if CAPI is not up for some reason.

This will make the whole thing a bit best-effort but I think that's acceptable.

**Comment by Ankitasw** [comment]
Is this issue up for grab? I would like to work on it if not started.

**Comment by killianmuldoon** [comment]
>Is this issue up for grab? I would like to work on it if not started.

I think you can feel free to assign yourself - thanks for taking a look at this!

**Comment by killianmuldoon** [comment]
@Ankitasw have you been able to make progress on this?

**Comment by killianmuldoon** [comment]
/unassign
/help

We should really try to get some solution on this as we move towards removing v1alpha4.

@Ankitasw in case you have more bandwidth now.

**Comment by sbueringer** [comment]
I'm not sure if we should really have a webhook on CRDs. I don't want to break mgmt clusters if CAPI is not up for this (almost nobody will run into this problem). There has also been no request for this since ClusterClass exists (CAPI 1.0 or even 0.4?)

@fabriziopandini wdyt?

**Comment by rexagod** [comment]
I see.

The latter concern makes sense since this adds guardrails for a rather rare event, if users are pinned to a non-storage version for longer than the provider expects (more likely to happen) or if the provider has unrealistic SLAs on non-storage version(s) EOL timelines and drops them at non-regular or erratic intervals (less likely to happen), in which case, the webhook will kick in to save the day.

The former one should be good though, since we set [`failurePolicy: Ignore`](https://github.com/kubernetes-sigs/cluster-api/pull/13954/changes#diff-7cd1951c7623560f1ee1e36775dc44f9146b58545e5f961e3c34a499ad67eb65R274) for the webhook, so CAPI being down shouldn't make the webhook blocking?

**Comment by sbueringer** [comment]
> since we set [failurePolicy: Ignore](https://github.com/kubernetes-sigs/cluster-api/pull/13954/changes#diff-7cd1951c7623560f1ee1e36775dc44f9146b58545e5f961e3c34a499ad67eb65R274) for the webhook, so CAPI being down shouldn't make the webhook blocking?

But Ignore makes the validation unreliable

**Comment by fabriziopandini** [comment]
I agree with @sbueringer 

I think that what we are discussing are upgrade pre-flight checks, and such, in my mind they should be part of the components responsible for upgrades and not implemented as web hooks. Also, the components responsible for upgrades can block an upgrade before it starts, while web hooks can only block in the middle of an upgrade, which is something that we should avoid.

Sort of orthogonal, after changes introduced in v1beta2, the problem is now scoped to ClusterClass only. Is this correct?

**Comment by sbueringer** [comment]
> Sort of orthogonal, after changes introduced in v1beta2, the problem is now scoped to ClusterClass only. Is this correct?

Yes, many of the apiVersion fields were replaced with apiGroup. Yes ClusterClass is still there, but also e.g. MHC remediationTemplate. Would require a full audit of our APIs for all places where APIVersion is used today

**Comment by rexagod** [comment]
> But Ignore makes the validation unreliable

Right, however, my understanding while patching this was that's [something we've already weighed](https://github.com/kubernetes-sigs/cluster-api/issues/8566#issuecomment-1544347484) (unless, as it occurs to me now, it's not valid anymore since it has been a while)? I was under the impression that we'd want this to be more of a subtle nudge to users, since this isn't a field CAPI reconciles.

But I do agree with @fabriziopandini's case, and I may be missing something here, but wouldn't it be better to ensure this is covered from both angles (upgrades as well as edits)? Also, I presume the precedent set here will apply to MHC and other resources that still rely on `apiVersion`?

**Comment by fabriziopandini** [comment]
TBH, If we really want to do something here, I would start covering the upgrade path which seems the one that will provide more benefits to users in the short term. Also I think that if we start considering edits to CRD we are going down a slippery path.

WRT to other resources that still rely on apiVersion, no objection to add more resources, but it only really makes sense if the referenced CRD are managed by the same upgrade precess

**Comment by rexagod** [comment]
> Also I think that if we start considering edits to CRD we are going down a slippery path.

Hard agree, I never considered that. I'll try to pivot this to follow the upgrade path, because that indeed forms a precedent that keeps CRD validations bounded.
